flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -269,22 +269,60 @@ case class Join(
     condition: Option[Expression]) extends BinaryNode {
 
   override def output: Seq[Attribute] = {
-    joinType match {
-      case JoinType.INNER => left.output ++ right.output
-      case j => throw new ValidationException(s"Unsupported JoinType: $j")
+    left.output ++ right.output
+  }
+
+  private case class JoinFieldReference(
+    name: String,
+    resultType: TypeInformation[_],
+    left: RelNode,
+    right: RelNode) extends Attribute {
+
+    override def toString = s"'$name"
+
+    override def toRexNode(implicit relBuilder: RelBuilder): RexNode = {
+      val joinInputField = if (left.getRowType.getFieldNames.contains(name)) {
+        val field = left.getRowType.getField(name, false, false)
+        (field.getIndex, field.getType)
+      } else {
+        val field = right.getRowType.getField(name, false, false)
+        (field.getIndex + left.getRowType.getFieldCount, field.getType)
+      }
+
+      new RexInputRef(joinInputField._1, joinInputField._2)
+    }
+
+    override def withName(newName: String): Attribute = {
+      if (newName == name) {
+        this
+      } else {
+        JoinFieldReference(newName, resultType, left, right)
+      }
     }
   }
 
   override protected[logical] def construct(relBuilder: RelBuilder): RelBuilder = {
-    joinType match {
-      case JoinType.INNER =>
-        left.construct(relBuilder)
-        right.construct(relBuilder)
-        relBuilder.join(JoinRelType.INNER,
-          condition.map(_.toRexNode(relBuilder)).getOrElse(relBuilder.literal(true)))
-      case _ =>
-        throw new ValidationException(s"Unsupported JoinType: $joinType")
+    left.construct(relBuilder)
+    right.construct(relBuilder)
+    val partialFunction: PartialFunction[Expression, Expression] = {
```

Comment:
I think the resolution into `JoinFieldReference` should happen in the `resolveExpression()` method / phase and not in `construct()`. See the docs in `LogicalNode`. What do you think @yjshen?